apps/cowswap-frontend/src/modules/fiatAmount/services/fetchCurrencyFiatPrice.ts

```diff
@@ -0,0 +1,49 @@
+import { Token } from '@uniswap/sdk-core'
+
+import { LONG_PRECISION } from 'legacy/constants'
+
+import {
+  COINGECKO_RATE_LIMIT_TIMEOUT,
+  CoingeckoRateLimitError,
+  getCoingeckoPrice,
+  UnsupporedCoingeckoPlatformError,
+} from '../apis/getCoingeckoPrice'
+import { getCowProtocolFiatPrice } from '../apis/getCowProtocolFiatPrice'
+
+let coingeckoRateLimitHitTimestamp: null | number = null
+
+/**
+ * Fetches fiat price for a given currency from coingecko or CowProtocol
+ * CowProtocol is used as a fallback
+ * When coingecko rate limit is hit, CowProtocol will be used for 1 minute
+ */
+export function fetchCurrencyFiatPrice(currency: Token, usdcPrice$: Promise<number | null>): Promise<number | null> {
+  const shouldSkipCoingecko =
+    !!coingeckoRateLimitHitTimestamp && Date.now() - coingeckoRateLimitHitTimestamp < COINGECKO_RATE_LIMIT_TIMEOUT
+
+  if (coingeckoRateLimitHitTimestamp && !shouldSkipCoingecko) {
+    coingeckoRateLimitHitTimestamp = null
+  }
+
+  const request = shouldSkipCoingecko
+    ? getCowProtocolFiatPrice(currency, usdcPrice$)
+    : getCoingeckoPrice(currency).catch((error) => {
+        if (error instanceof CoingeckoRateLimitError) {
+          coingeckoRateLimitHitTimestamp = Date.now()
+          console.error('Coingecko request limit reached')
+        } else if (!(error instanceof UnsupporedCoingeckoPlatformError)) {
+          console.error('Cannot fetch coingecko price', error)
+        }
```

Comment:
While this makes this part of the code cleaner, I don't like to do flow control based on exceptions.
I mean, when on goerli we know it'll fail and wait for the exception to use the fallback.

A still clean way can be to include the network filter with the suggested helper fn above and avoid calling coigecko when we know it's not supported.

Reply:
Makes sense, thanks!
Refactored